open-adventure: init at 1.20

EmanuelM153 opened the pull request:
[open-adventure](http://www.catb.org/~esr/open-adventure/history.html) is a forward-port of the Crowther/Woods Adventure 2.5 game from 1995

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

EmanuelM153 committed:
maintainers: add EmanuelM153

EmanuelM153 committed:
open-adventure: init at 1.20

SigmaSquadron commented on pkgs/by-name/op/open-adventure/package.nix:
```diff
@@ -0,0 +1,56 @@
+  lib,
+  stdenv,
+  fetchFromGitLab,
+  python312,
```
```suggestion
```

You want `python312Packages.python` instead.

SigmaSquadron commented on pkgs/by-name/op/open-adventure/package.nix:
```diff
@@ -0,0 +1,56 @@
+  src = fetchFromGitLab {
+    owner = "esr";
+    repo = "open-adventure";
+    rev = "refs/tags/${finalAttrs.version}";
```
```suggestion
    tag = finalAttrs.version;
```

SigmaSquadron commented on pkgs/by-name/op/open-adventure/package.nix:
```diff
@@ -0,0 +1,56 @@
+
+  buildInputs = [ libedit ];
+
+  buildPhase = ''
+    patchShebangs make_dungeon.py
+    make
+  '';
```
```suggestion
  postPatch = ''
    patchShebangs --build make_dungeon.py
  '';
```

Avoid overriding `buildPhase`.

SigmaSquadron commented on pkgs/by-name/op/open-adventure/package.nix:
```diff
@@ -0,0 +1,56 @@
+  ];
+
+  buildInputs = [ libedit ];
+
```
```suggestion

  doCheck = true;

```

Looks like upstream has included some tests in the Makefile. Would be great if we also built those.

SigmaSquadron commented on pkgs/by-name/op/open-adventure/package.nix:
```diff
@@ -0,0 +1,56 @@
+    runHook postInstall
+  '';
+
+  meta = with lib; {
```
```suggestion
  meta = {
```
Nested `with;` expressions are considered a bad practice.

SigmaSquadron commented on pkgs/by-name/op/open-adventure/package.nix:
```diff
@@ -0,0 +1,56 @@
+  stdenv,
+  fetchFromGitLab,
+  python312,
+  python312Packages,
```
```suggestion
  python3Packages,
```
Are you sure this doesn't work with an unpinned Python?

SigmaSquadron reviewed (commented):
Welcome to Nixpkgs! Thanks for contributing this game.

EmanuelM153 commented:
@SigmaSquadron Thank you for your corrections! I've followed your suggestions and added the upstream tests.
Please let me know if there are any other things to improve 😉.

Btw, I am not quite sure if the commands I added under the `preCheck` attribute are the best solution.
Maybe the problem with cppcheck is solved using an older version, and I don't know if using the `substituteInPlace` command, to replace /bin/echo, is the best approach.

SigmaSquadron commented on pkgs/by-name/op/open-adventure/package.nix:
```diff
@@ -0,0 +1,71 @@
+    substituteInPlace Makefile --replace-fail "--template " "--template="
+
+    # Replace /bin/echo with the actual path in the nix store
+    substituteInPlace tests/tapview --replace-fail "/bin/echo" ${coreutils}/bin/echo
```
```suggestion
    substituteInPlace tests/tapview --replace-fail "/bin/echo" ${lib.getExe' coreutils "echo"}
```

SigmaSquadron reviewed (commented):
`substituteInPlace` was indeed the right call. It's pretty normal to substitute FHS paths with Nix paths in `patchPhase`.

SigmaSquadron commented on pkgs/by-name/op/open-adventure/package.nix:
```diff
@@ -0,0 +1,71 @@
+
+  preCheck = ''
+    # Fix the lack of the = symbol in the cli of cppcheck
+    substituteInPlace Makefile --replace-fail "--template " "--template="
```
Would it be possible to upstream this fix?

SigmaSquadron commented on pkgs/by-name/op/open-adventure/package.nix:
```diff
@@ -0,0 +1,71 @@
+    cppcheck
+  ];
+
+  preCheck = ''
```
```suggestion
  postPatch = ''
```
This can be merged with `postPatch`.

ethancedwards8 commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 370955`

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>open-adventure</li>
  </ul>
</details>

lucasew commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 370955`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>open-adventure</li>
  </ul>
</details>

EmanuelM153 commented:
@SigmaSquadron I have followed your corrections, and I had also created an issue on upstream to apply the fix, however they haven't answered. Furthermore, I was reviewing other issues on their [repository](https://gitlab.com/esr/open-adventure) and lately they are taking a long time to solve them (even some weeks), probably because it is not a project under development. So, maybe the package can be added now (I'll be paying attention to the appliance of the fix, and will update the package accordingly).

Btw, I don't know if the commits should be squashed 😅.

Regards,
EmanuelM153

SigmaSquadron commented:
> Btw, I don't know if the commits should be squashed 😅.

When in doubt, squash! This PR should have two commits total.

SigmaSquadron commented on pkgs/by-name/op/open-adventure/package.nix:
```diff
@@ -0,0 +1,84 @@
+
+  nativeBuildInputs = [
+    python3Packages.python
+    python3Packages.pyyaml
```
```suggestion
```
`pyyaml` isn't a program that can be executed, so it goes to `buildInputs`.

SigmaSquadron commented on pkgs/by-name/op/open-adventure/package.nix:
```diff
@@ -0,0 +1,84 @@
+    substituteInPlace tests/tapview --replace-fail "/bin/echo" ${lib.getExe' coreutils "echo"}
+  '';
+
+  postBuild = ''
+    make advent.6
+  '';
```
Please use [`buildFlags`](https://nixos.org/manual/nixpkgs/unstable/#var-stdenv-buildFlags) instead.

SigmaSquadron commented on pkgs/by-name/op/open-adventure/package.nix:
```diff
@@ -0,0 +1,84 @@
+    # Fix the lack of the = symbol in the cli of cppcheck
+    substituteInPlace Makefile --replace-fail "--template " "--template="
+
+    # Replace /bin/echo with the actual path in the nix store
```
```suggestion
```
I'm a fan of [superfluous comments](https://github.com/NixOS/nixpkgs/pull/331867#discussion_r1704685492) myself, but this is a pretty well-known pattern in Nixpkgs, and doesn't really need to be there. If you feel strongly about it, feel free to leave it here; this isn't a blocking comment.

SigmaSquadron commented on pkgs/by-name/op/open-adventure/package.nix:
```diff
@@ -0,0 +1,84 @@
+    make advent.6
+  '';
+
+  preInstall = ''
+    mkdir -vp "$out/bin" "$out/share/man/man6" "$out/share/applications/" "$out/share/icons/hicolor/scalable/apps"
+  '';
+
+  installPhase = ''
+    runHook preInstall
+    cp ./advent $out/bin
+    runHook postInstall
+  '';
+
+  postInstall = ''
+    cp ./advent.6 $out/share/man/man6
+    cp ./advent.desktop $out/share/applications
+    cp ./advent.svg $out/share/icons/hicolor/scalable/apps
+  '';
```
```suggestion
  installPhase = ''
    runHook preInstall
    
    mkdir -vp "$out/bin" "$out/share/man/man6" "$out/share/applications/" "$out/share/icons/hicolor/scalable/apps"
    cp ./advent $out/bin
    cp ./advent.6 $out/share/man/man6
    cp ./advent.desktop $out/share/applications
    cp ./advent.svg $out/share/icons/hicolor/scalable/apps
    
    runHook postInstall
  '';
```

This should also use the `install` command whenever possible. I don't know the package filesystem layout to suggest an implementation.

EmanuelM153 commented:
Thank you for your suggestions, I've applied them now

SigmaSquadron commented on pkgs/by-name/op/open-adventure/package.nix:
```diff
@@ -0,0 +1,81 @@
+    substituteInPlace tests/tapview --replace-fail "/bin/echo" ${lib.getExe' coreutils "echo"}
+  '';
+
+  buildFlags = ''
+    advent.6
+  '';
```
```suggestion
  buildFlags = [
    "advent.6"
  ];
```
`buildFlags` is a list.

EmanuelM153 replied:
Ahh ok ok

EmanuelM153 commented:
Ahh ok thank you.
I also forgot the use of the `install` command. It is now updated

SigmaSquadron reviewed (approved):
```console
$ ./results/open-adventure-x86_64-linux/bin/advent

Welcome to Adventure!!  Would you like instructions?

```

---

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 370955`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>open-adventure</li>
  </ul>
</details>

SigmaSquadron commented:
Gotta say, pretty fun game :)

EmanuelM153 commented:
Jsjsj yeah. Thank you for guiding me in this process :)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2242

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-in-distress/3604/98

TomaSajt commented on pkgs/by-name/op/open-adventure/package.nix:
```diff
@@ -0,0 +1,81 @@
+    # https://gitlab.com/esr/open-adventure/-/issues/70
+    substituteInPlace Makefile --replace-fail "--template " "--template="
+
+    substituteInPlace tests/tapview --replace-fail "/bin/echo" ${lib.getExe' coreutils "echo"}
```
Does this only run during testing? If yes, I just recommend patching it to just be `echo`, since we know it's on our PATH.

TomaSajt commented on pkgs/by-name/op/open-adventure/package.nix:
```diff
@@ -0,0 +1,81 @@
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -vp "$out/bin" "$out/share/man/man6" "$out/share/applications/" "$out/share/icons/hicolor/scalable/apps"
+    install -m 555 ./advent $out/bin
+    install -m 444 ./advent.6 $out/share/man/man6
+    install -m 444 ./advent.desktop $out/share/applications
+    install -m 444 ./advent.svg $out/share/icons/hicolor/scalable/apps
```
A possible alternative that doesn't need mkdir. You don't need to do this if you don't want to.
```suggestion
    install -Dm555 ./advent -t $out/bin
    install -Dm444 ./advent.6 -t $out/share/man/man6
    install -Dm444 ./advent.desktop -t $out/share/applications
    install -Dm444 ./advent.svg -t $out/share/icons/hicolor/scalable/apps
```

Also, man-pages should ideally be installed via `installManPage` from the `installShellFiles` helper script collection. IMO this is not a hard requirement, so you can keep it as is.

TomaSajt commented on pkgs/by-name/op/open-adventure/package.nix:
```diff
@@ -0,0 +1,81 @@
+    python3Packages.pyyaml
+    libedit
+  ];
+
```
consider setting `strictDeps = true;` to ensure that you've added your deps to the correct place (native vs non-native inputs). 
```suggestion

  strictDeps = true;

```

TomaSajt reviewed (commented):
The changes look fine, left some possible improvement ideas.

TomaSajt commented:
wow I just missed the merge by a minute :) Oh well, the improvements can still be made if you want to,

EmanuelM153 commented:
Thanks! I will take into account this improvements in my future contributions :)
